src/gui_common/GUICommon.cs

```diff
@@ -55,6 +57,18 @@ public static void PopupMinSizeMarginPosition(Popup popup)
         popup.RectPosition = new Vector2(left, top);
     }
 
+    public static void SmoothlyUpdateBar(TextureProgress bar, float target, float delta)
+    {
+        if (delta <= 0.0f)
+        {
+            var paramName = "delta must be positive!";
+            throw new ArgumentOutOfRangeException(paramName);
+        }
```

Comment:
```suggestion
        // https://github.com/Revolutionary-Games/Thrive/issues/1976
        if (delta <= 0)
            return;
```

Reply:
Now that this is a public function, I don't think this is a good idea. Future developers will be free to use this code in any UI code, including outside of update methods where "delta" becomes a less obvious concept. Giving a default value of 0.0f, or worse SOMETIMES passing in that value, will result in one of those implicit quiet failures that leads to those hair-pulling kinds of bugs. With code in a helper library you can't make assumptions about how it will be used.

Reply:
The game needs to **not** crash in any GUI functionality because of the silliness done by Godot engine.
Either this code gets added here, or the biggest warning imaginable needs to be added as a documentation comment on this method and all usages of this method that take in a delta value from Godot need to be verified (and preferrably also added a comment there) that negative delta can't be passed in.

Wouldn't this interpolation not do anything anyway passing in a 0 delta? So anyway this method doesn't work if proper delta from Godot is not passed in. An xml doc comment saying that you *must* pass in a real delta otherwise the method won't do anything, would be enough, right?

Reply:
Lets face it: you and I don't read all the comments on the helper libraries before we use them, and we can't expect the next guys to do so. More broadly though, I think that we can do better than just saying, "hey, if you use this generic code in a certain way it will do strange stuff." Right now this procedure still makes assumptions about the context in which it's called, which is a clear example of how NOT to do the Single Responsibility Principle. 

I can think of two solutions. The lazy one is to just put this back in MicrobeHUD and not claim it's generic. The more exciting idea would be to replace the delta with a speed parameter that helps determine how fast the bar moves. MicrobeHUD could then pick a speed based on delta within it's on context, and another user can pick speed whatever way they want. I won't shed a tear if a dev sets speed to zero, then wonders why the bar doesn't move.

Reply:
After sleeping on this, I came up with a pretty obvious solution: use GD.PrintErr here to output the error message in red. That way it is noticeable, but will not crash the game.

In general I agree with you that it would be nice to throw an exception if you call a method with bogus data. However, games are a case where you want to avoid crashing much more than in other software. We already have places in the code that get a serious error, but just print that and continue, because it's much better if some part of the game doesn't work than making the game crash on the player.

If it helps you think about this as GUI specific code, then why not rename this to explicitly include the fact that this uses elapsed time to move the bar? That way it is obvious from the name as well that this needs a real elapsed time to work.

Reply:
Well while you were sleeping I tried out removing delta from the function altogether: https://github.com/Revolutionary-Games/Thrive/pull/2333/files#diff-d5f5074b2176ba208d7d0381045c36e02b673dd709eb00647a93e9187d15aae9R66

I'm happy to compromise with the PrintErr, but now that I've done it I kind of like the speed param. Let me know what you think.

Reply:
I think it's better to have the delta in the generic place, I reckon that it is way more common to use it from GUI code that does now the actual delta, meaning that the logic for going to delta to the speed parameter gets duplicated everywhere, which is not good. The reason why delta is used in the first place is so that the bar move speed is tied to real time instead of framerate, as otherwise at 400 FPS vs 60 FPS the GUI animations will be out of wack.